{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tensorflow.compat.v1 import ConfigProto\n",
    "from tensorflow.compat.v1 import InteractiveSession\n",
    "import os\n",
    "import numpy as np\n",
    "import tensorflow as tf\n",
    "from tensorflow.keras.preprocessing.image import ImageDataGenerator\n",
    "from tensorflow.keras.layers import Conv2D\n",
    "from tensorflow.keras.layers import Dense\n",
    "from tensorflow.keras.regularizers import l2\n",
    "from tensorflow.keras.preprocessing import image\n",
    "import torch\n",
    "from PIL import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "config = ConfigProto()\n",
    "config.gpu_options.per_process_gpu_memory_fraction = 0.8\n",
    "config.gpu_options.allow_growth = True\n",
    "session = InteractiveSession(config=config)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 134271 images belonging to 38 classes.\n",
      "Found 32955 images belonging to 38 classes.\n"
     ]
    }
   ],
   "source": [
    "data_dir = \"archive/\"\n",
    "train_dir = data_dir + \"/train\"\n",
    "valid_dir = data_dir + \"/valid\"\n",
    "train_datagen = ImageDataGenerator(rescale = 1./255)\n",
    "training_set = train_datagen.flow_from_directory(train_dir,\n",
    "                                    target_size = (64, 64),\n",
    "                                    batch_size = 32,\n",
    "                                    class_mode = 'categorical')\n",
    "valid_datagen = ImageDataGenerator(rescale = 1./255)\n",
    "valid_set = valid_datagen.flow_from_directory(valid_dir,\n",
    "                                            target_size = (64, 64),\n",
    "                                            batch_size = 32,\n",
    "                                            class_mode = 'categorical')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Apple___Apple_scab',\n",
       " 'Apple___Black_rot',\n",
       " 'Apple___Cedar_apple_rust',\n",
       " 'Apple___healthy',\n",
       " 'Blueberry___healthy',\n",
       " 'Cherry_(including_sour)___healthy',\n",
       " 'Cherry_(including_sour)___Powdery_mildew',\n",
       " 'Corn_(maize)___Cercospora_leaf_spot_Gray_leaf_spot',\n",
       " 'Corn_(maize)___Common_rust_',\n",
       " 'Corn_(maize)___healthy',\n",
       " 'Corn_(maize)___Northern_Leaf_Blight',\n",
       " 'Grape___Black_rot',\n",
       " 'Grape___Esca_(Black_Measles)',\n",
       " 'Grape___healthy',\n",
       " 'Grape___Leaf_blight_(Isariopsis_Leaf_Spot)',\n",
       " 'Orange___Haunglongbing_(Citrus_greening)',\n",
       " 'Peach___Bacterial_spot',\n",
       " 'Peach___healthy',\n",
       " 'Pepper,_bell___Bacterial_spot',\n",
       " 'Pepper,_bell___healthy',\n",
       " 'Potato___Early_blight',\n",
       " 'Potato___healthy',\n",
       " 'Potato___Late_blight',\n",
       " 'Raspberry___healthy',\n",
       " 'Soybean___healthy',\n",
       " 'Squash___Powdery_mildew',\n",
       " 'Strawberry___healthy',\n",
       " 'Strawberry___Leaf_scorch',\n",
       " 'Tomato___Bacterial_spot',\n",
       " 'Tomato___Early_blight',\n",
       " 'Tomato___healthy',\n",
       " 'Tomato___Late_blight',\n",
       " 'Tomato___Leaf_Mold',\n",
       " 'Tomato___Septoria_leaf_spot',\n",
       " 'Tomato___Spider_mites_Two-spotted_spider_mite',\n",
       " 'Tomato___Target_Spot',\n",
       " 'Tomato___Tomato_mosaic_virus',\n",
       " 'Tomato___Tomato_Yellow_Leaf_Curl_Virus']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diseases =os.listdir(train_dir)\n",
    "diseases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "cnn = tf.keras.models.Sequential()\n",
    "cnn.add(tf.keras.layers.Conv2D(filters=32,padding=\"same\",kernel_size=3, activation='relu', strides=2, input_shape=[64, 64, 3]))\n",
    "cnn.add(tf.keras.layers.MaxPool2D(pool_size=2, strides=2))\n",
    "cnn.add(tf.keras.layers.Conv2D(filters=32,padding='same',kernel_size=3, activation='relu'))\n",
    "cnn.add(tf.keras.layers.MaxPool2D(pool_size=2, strides=2))\n",
    "cnn.add(tf.keras.layers.Flatten())\n",
    "cnn.add(tf.keras.layers.Dense(units=128, activation='relu'))\n",
    "cnn.add(Dense(38, kernel_regularizer=tf.keras.regularizers.l2(0.01),activation ='softmax'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/10\n"
     ]
    },
    {
     "ename": "UnimplementedError",
     "evalue": "Graph execution error:\n\nDetected at node 'sequential/conv2d/Relu' defined at (most recent call last):\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\runpy.py\", line 197, in _run_module_as_main\n      return _run_code(code, main_globals, None,\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\runpy.py\", line 87, in _run_code\n      exec(code, run_globals)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\ipykernel_launcher.py\", line 16, in <module>\n      app.launch_new_instance()\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\traitlets\\config\\application.py\", line 846, in launch_instance\n      app.start()\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\ipykernel\\kernelapp.py\", line 677, in start\n      self.io_loop.start()\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\tornado\\platform\\asyncio.py\", line 199, in start\n      self.asyncio_loop.run_forever()\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\asyncio\\base_events.py\", line 601, in run_forever\n      self._run_once()\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\asyncio\\base_events.py\", line 1905, in _run_once\n      handle._run()\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\asyncio\\events.py\", line 80, in _run\n      self._context.run(self._callback, *self._args)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 471, in dispatch_queue\n      await self.process_one()\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 460, in process_one\n      await dispatch(*args)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 367, in dispatch_shell\n      await result\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 662, in execute_request\n      reply_content = await reply_content\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\ipykernel\\ipkernel.py\", line 360, in do_execute\n      res = shell.run_cell(code, store_history=store_history, silent=silent)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\ipykernel\\zmqshell.py\", line 532, in run_cell\n      return super().run_cell(*args, **kwargs)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2863, in run_cell\n      result = self._run_cell(\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2909, in _run_cell\n      return runner(coro)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\IPython\\core\\async_helpers.py\", line 129, in _pseudo_sync_runner\n      coro.send(None)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3106, in run_cell_async\n      has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3309, in run_ast_nodes\n      if await self.run_code(code, result, async_=asy):\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3369, in run_code\n      exec(code_obj, self.user_global_ns, self.user_ns)\n    File \"C:\\Users\\ektap\\AppData\\Local\\Temp\\ipykernel_12964\\396496758.py\", line 2, in <cell line: 2>\n      r=cnn.fit(x = training_set, validation_data = valid_set, epochs = 10)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\utils\\traceback_utils.py\", line 65, in error_handler\n      return fn(*args, **kwargs)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\training.py\", line 1564, in fit\n      tmp_logs = self.train_function(iterator)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\training.py\", line 1160, in train_function\n      return step_function(self, iterator)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\training.py\", line 1146, in step_function\n      outputs = model.distribute_strategy.run(run_step, args=(data,))\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\training.py\", line 1135, in run_step\n      outputs = model.train_step(data)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\training.py\", line 993, in train_step\n      y_pred = self(x, training=True)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\utils\\traceback_utils.py\", line 65, in error_handler\n      return fn(*args, **kwargs)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\training.py\", line 557, in __call__\n      return super().__call__(*args, **kwargs)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\utils\\traceback_utils.py\", line 65, in error_handler\n      return fn(*args, **kwargs)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\base_layer.py\", line 1097, in __call__\n      outputs = call_fn(inputs, *args, **kwargs)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\utils\\traceback_utils.py\", line 96, in error_handler\n      return fn(*args, **kwargs)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\sequential.py\", line 410, in call\n      return super().call(inputs, training=training, mask=mask)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\functional.py\", line 510, in call\n      return self._run_internal_graph(inputs, training=training, mask=mask)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\functional.py\", line 667, in _run_internal_graph\n      outputs = node.layer(*args, **kwargs)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\utils\\traceback_utils.py\", line 65, in error_handler\n      return fn(*args, **kwargs)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\base_layer.py\", line 1097, in __call__\n      outputs = call_fn(inputs, *args, **kwargs)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\utils\\traceback_utils.py\", line 96, in error_handler\n      return fn(*args, **kwargs)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\layers\\convolutional\\base_conv.py\", line 314, in call\n      return self.activation(outputs)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\activations.py\", line 317, in relu\n      return backend.relu(\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\backend.py\", line 5366, in relu\n      x = tf.nn.relu(x)\nNode: 'sequential/conv2d/Relu'\nDNN library is not found.\n\t [[{{node sequential/conv2d/Relu}}]] [Op:__inference_train_function_905]",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mUnimplementedError\u001b[0m                        Traceback (most recent call last)",
      "Input \u001b[1;32mIn [6]\u001b[0m, in \u001b[0;36m<cell line: 2>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      1\u001b[0m cnn\u001b[38;5;241m.\u001b[39mcompile(optimizer \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124madam\u001b[39m\u001b[38;5;124m'\u001b[39m, loss \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124msquared_hinge\u001b[39m\u001b[38;5;124m'\u001b[39m, metrics \u001b[38;5;241m=\u001b[39m [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124maccuracy\u001b[39m\u001b[38;5;124m'\u001b[39m])\n\u001b[1;32m----> 2\u001b[0m r\u001b[38;5;241m=\u001b[39m\u001b[43mcnn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfit\u001b[49m\u001b[43m(\u001b[49m\u001b[43mx\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mtraining_set\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mvalidation_data\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mvalid_set\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mepochs\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m10\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32m~\\anaconda3\\lib\\site-packages\\keras\\utils\\traceback_utils.py:70\u001b[0m, in \u001b[0;36mfilter_traceback.<locals>.error_handler\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m     67\u001b[0m     filtered_tb \u001b[38;5;241m=\u001b[39m _process_traceback_frames(e\u001b[38;5;241m.\u001b[39m__traceback__)\n\u001b[0;32m     68\u001b[0m     \u001b[38;5;66;03m# To get the full stack trace, call:\u001b[39;00m\n\u001b[0;32m     69\u001b[0m     \u001b[38;5;66;03m# `tf.debugging.disable_traceback_filtering()`\u001b[39;00m\n\u001b[1;32m---> 70\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m e\u001b[38;5;241m.\u001b[39mwith_traceback(filtered_tb) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;28mNone\u001b[39m\n\u001b[0;32m     71\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[0;32m     72\u001b[0m     \u001b[38;5;28;01mdel\u001b[39;00m filtered_tb\n",
      "File \u001b[1;32m~\\anaconda3\\lib\\site-packages\\tensorflow\\python\\eager\\execute.py:54\u001b[0m, in \u001b[0;36mquick_execute\u001b[1;34m(op_name, num_outputs, inputs, attrs, ctx, name)\u001b[0m\n\u001b[0;32m     52\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m     53\u001b[0m   ctx\u001b[38;5;241m.\u001b[39mensure_initialized()\n\u001b[1;32m---> 54\u001b[0m   tensors \u001b[38;5;241m=\u001b[39m pywrap_tfe\u001b[38;5;241m.\u001b[39mTFE_Py_Execute(ctx\u001b[38;5;241m.\u001b[39m_handle, device_name, op_name,\n\u001b[0;32m     55\u001b[0m                                       inputs, attrs, num_outputs)\n\u001b[0;32m     56\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m core\u001b[38;5;241m.\u001b[39m_NotOkStatusException \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[0;32m     57\u001b[0m   \u001b[38;5;28;01mif\u001b[39;00m name \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "\u001b[1;31mUnimplementedError\u001b[0m: Graph execution error:\n\nDetected at node 'sequential/conv2d/Relu' defined at (most recent call last):\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\runpy.py\", line 197, in _run_module_as_main\n      return _run_code(code, main_globals, None,\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\runpy.py\", line 87, in _run_code\n      exec(code, run_globals)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\ipykernel_launcher.py\", line 16, in <module>\n      app.launch_new_instance()\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\traitlets\\config\\application.py\", line 846, in launch_instance\n      app.start()\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\ipykernel\\kernelapp.py\", line 677, in start\n      self.io_loop.start()\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\tornado\\platform\\asyncio.py\", line 199, in start\n      self.asyncio_loop.run_forever()\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\asyncio\\base_events.py\", line 601, in run_forever\n      self._run_once()\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\asyncio\\base_events.py\", line 1905, in _run_once\n      handle._run()\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\asyncio\\events.py\", line 80, in _run\n      self._context.run(self._callback, *self._args)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 471, in dispatch_queue\n      await self.process_one()\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 460, in process_one\n      await dispatch(*args)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 367, in dispatch_shell\n      await result\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 662, in execute_request\n      reply_content = await reply_content\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\ipykernel\\ipkernel.py\", line 360, in do_execute\n      res = shell.run_cell(code, store_history=store_history, silent=silent)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\ipykernel\\zmqshell.py\", line 532, in run_cell\n      return super().run_cell(*args, **kwargs)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2863, in run_cell\n      result = self._run_cell(\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2909, in _run_cell\n      return runner(coro)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\IPython\\core\\async_helpers.py\", line 129, in _pseudo_sync_runner\n      coro.send(None)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3106, in run_cell_async\n      has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3309, in run_ast_nodes\n      if await self.run_code(code, result, async_=asy):\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3369, in run_code\n      exec(code_obj, self.user_global_ns, self.user_ns)\n    File \"C:\\Users\\ektap\\AppData\\Local\\Temp\\ipykernel_12964\\396496758.py\", line 2, in <cell line: 2>\n      r=cnn.fit(x = training_set, validation_data = valid_set, epochs = 10)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\utils\\traceback_utils.py\", line 65, in error_handler\n      return fn(*args, **kwargs)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\training.py\", line 1564, in fit\n      tmp_logs = self.train_function(iterator)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\training.py\", line 1160, in train_function\n      return step_function(self, iterator)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\training.py\", line 1146, in step_function\n      outputs = model.distribute_strategy.run(run_step, args=(data,))\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\training.py\", line 1135, in run_step\n      outputs = model.train_step(data)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\training.py\", line 993, in train_step\n      y_pred = self(x, training=True)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\utils\\traceback_utils.py\", line 65, in error_handler\n      return fn(*args, **kwargs)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\training.py\", line 557, in __call__\n      return super().__call__(*args, **kwargs)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\utils\\traceback_utils.py\", line 65, in error_handler\n      return fn(*args, **kwargs)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\base_layer.py\", line 1097, in __call__\n      outputs = call_fn(inputs, *args, **kwargs)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\utils\\traceback_utils.py\", line 96, in error_handler\n      return fn(*args, **kwargs)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\sequential.py\", line 410, in call\n      return super().call(inputs, training=training, mask=mask)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\functional.py\", line 510, in call\n      return self._run_internal_graph(inputs, training=training, mask=mask)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\functional.py\", line 667, in _run_internal_graph\n      outputs = node.layer(*args, **kwargs)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\utils\\traceback_utils.py\", line 65, in error_handler\n      return fn(*args, **kwargs)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\base_layer.py\", line 1097, in __call__\n      outputs = call_fn(inputs, *args, **kwargs)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\utils\\traceback_utils.py\", line 96, in error_handler\n      return fn(*args, **kwargs)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\layers\\convolutional\\base_conv.py\", line 314, in call\n      return self.activation(outputs)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\activations.py\", line 317, in relu\n      return backend.relu(\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\backend.py\", line 5366, in relu\n      x = tf.nn.relu(x)\nNode: 'sequential/conv2d/Relu'\nDNN library is not found.\n\t [[{{node sequential/conv2d/Relu}}]] [Op:__inference_train_function_905]"
     ]
    }
   ],
   "source": [
    "cnn.compile(optimizer = 'adam', loss = 'squared_hinge', metrics = ['accuracy'])\n",
    "r=cnn.fit(x = training_set, validation_data = valid_set, epochs = 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model: \"sequential\"\n",
      "_________________________________________________________________\n",
      " Layer (type)                Output Shape              Param #   \n",
      "=================================================================\n",
      " conv2d (Conv2D)             (None, 32, 32, 32)        896       \n",
      "                                                                 \n",
      " max_pooling2d (MaxPooling2D  (None, 16, 16, 32)       0         \n",
      " )                                                               \n",
      "                                                                 \n",
      " conv2d_1 (Conv2D)           (None, 16, 16, 32)        9248      \n",
      "                                                                 \n",
      " max_pooling2d_1 (MaxPooling  (None, 8, 8, 32)         0         \n",
      " 2D)                                                             \n",
      "                                                                 \n",
      " flatten (Flatten)           (None, 2048)              0         \n",
      "                                                                 \n",
      " dense (Dense)               (None, 128)               262272    \n",
      "                                                                 \n",
      " dense_1 (Dense)             (None, 38)                4902      \n",
      "                                                                 \n",
      "=================================================================\n",
      "Total params: 277,318\n",
      "Trainable params: 277,318\n",
      "Non-trainable params: 0\n",
      "_________________________________________________________________\n"
     ]
    }
   ],
   "source": [
    "cnn.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "plt.plot(r.history['loss'], label='train loss')\n",
    "plt.plot(r.history['val_loss'], label='val loss')\n",
    "plt.legend()\n",
    "plt.show()\n",
    "plt.savefig('LossVal_loss')\n",
    "\n",
    "# plot the accuracy\n",
    "plt.plot(r.history['accuracy'], label='train acc')\n",
    "plt.plot(r.history['val_accuracy'], label='val acc')\n",
    "plt.legend()\n",
    "plt.show()\n",
    "plt.savefig('AccVal_acc')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAEEAAABBCAIAAAABlV4SAAAh+UlEQVR4nEV6WYym2XnW2c/51n+rtau6epmlPTO2M/HMOGDscWKHELJZJCQSEhe5CCIBhQtuiARICQqRUAhC3ICElCAuI2TEhRMpkXGMHWMnzsQznsz0TM90d1XX/q/ffvbDxZk0N6XS/3/11Tnv8rzL88Df/PXf6LoOAOCcczYURaG1Ho1G1tq2bRFC1tokzyCE3nsAANKhMT0RKDhPPCHUW2tA4NPp2BuLEEIAWssCMNp0ecFBoJxzpQalhyzLHCRaawhhURRd1xFCsixbr9cAAM65MQZCCCHEGBNCrLV5njvnIIRSSillmqaEEAhh3/fee845QohcXFzs7u4qpZxzaZoihIQQCKG+75MkQQhJKb338UVCiEHW29vbiHLTDULwAIy1uusG5wHARHpHCA64dc4BDHpnirxYNw0AgIpicIARoLVmjFlrAQBSyqZp0jQdhgEAEEIYjUZVVVFKGWPeeymlMSZJEudcvFX8Q855tKnWGm1vb2utKaUIoWEYhmFQSg3DgDHGGIcQokkIISGEEAJPSF23VlqgVVtdeaUICATKPExuz56bif1u5WejPYYyAjKORxjpsuAYOqMGTijGOFqq73sAQF3Xo9EIQkgIEUJQSkMISZK0bds0DQAAYyyEcM6FELTW1lqMcXzeOWeMyfOcaK2999ZaSqngNE3TqqoAAAAApVSWZXVdF2zEGJNSWmvzNEEEWz1AN786OV0s1kmSUG4+96ULmojDI/ZTP3PnvfdHd17/5Lf/7BvWthfni6bp3n/v4adf+1tBtorQruvG4zGlVEq5t7cXPZwkSfS8tXYYhjzPQwjGGIxxNDHn3DnHOY8R3jRNnueUUmMMcc4xxrquy/N8sVj0Q8sYA9A7b6SUxqrRuLi8uLxz587ZpqJT3HkIgEEYoGTruU/duWPt5eWllPL3/yR78603zs8e3Lw5++yLR8/eKe48d4vufeZi/kY5nf3wF+4JTi8uzrAYp2m6WCySJKGUNk1TlqUxhlKaZRnnPISw2WySJIEQIoTquo7h4L1HCMUQ8t53XYcxttYihAiE0Dm3u7t7eXkZkykGmfc+TVMhxDAMd+/ebZpmZ2cHQmiMBcAnjFOKMcZ9308mkxDCk7OLT/7Acx+7dzM4+wd/+o2fLF/95OdvAOFm2zfXi2U3uMl4xGhtQ6CUjsdjhBCEMM/zYRhGo5FSarVaUUoppWVZRox5emhKKYSwaZphGGJ4Z1kWA77vexKz1jmXZVmSJMYY731MuzRNpZRFUVhrGWNKKa11kmRKDRjjpmmqaj0ejzEmlFLCb3jvMSDAQ5oV//tbf/z6j38uR+1kuv3hBw/HZXFxvaRJroc2hKCUEkL0fV8UBcZ4Pp8fHh7GeI4QhDHmnCdJslwuGWPxw8lkIqWMET4ejwEAXdfNZjMEAIj55L2Pj8bgQwgRQiKAEEIQQmmaMsZCCGVZ1nXtnKOMSDVABDbVGkKktaEMYeLvbB1QINbrISvKYpRNtkYeWuMkIqEoirquOeec86IoEEJJkty6dSv6Of4jIURMyOVyGRFFSokQ0lpjjLe2tkIITdPE94QQEIIkeBg85Cxp2zbC0fb2Nka0a4c8KwVPlVIAgKqqRqOR0l3XN0VRjEYTijiBLOV5wjKvTUKZUcY5b/FKG3J1uQFIpQkFIGIGt9oBAG7cuIExruuaEKK1rqqqaRpK6fHxccQiY0wIAQAwmUxiaFBKMcZKKYRQVVXGGMbYaDSq67rve0Qp3dvbiygBISzL0jm3Xq8ppcMwrFariGjxjYvFoizLWECi3wAAZ2dnSqk0TQEAlFLn3HQyy7Py7bffwYgihKJdVqvVZrNZr9d931trOedd1yml9vf3QwgIoclkMhqNYmZzzq21WutY9UIIQogIryGEmMPr9TpNU+ccicmQ57n3HhO4WCxiaQQB5Xmutb64uDg4OIj16Okr2rbN81yIbBiGoigghNFXTdNwzqXUy+V6u6GEsOjboijGZZlw0Q39ZrOZzWYhhBg5dV0nSRJCiNgdf482imXXe6+UikFOKdVa931/eHhojOm6znuP8jyPWRX9tb+/DwAoyzJWQSHE7u6utTZeFSHUdV0MZeccxpgxBgBACJVlGQs5AABjSggTIjHGjcfjeJphGM7OzqbTaTyWcy5GgjFGa62UKssyy7L5fO69j1kHIXzqiohjV1dXIYQbN25Ef0YroPVmiTAQCRtk1zRN27be+6ZpAoIBweVmzdPEQ6CdBRgV49F4PI5oDSFURrd9p4yGGCmlose7rqu7DRP09jN3IcIp4wRAGALGmCci4n3btsYYjCGl2FrdNBWEQanBWl2WeZLwrmvW66UQDGNojJrNJvHJw8MbEIauayjFGMOmqdbrJYpnisZ42sZEJ8YciPX8aRqcn5+HENI05ZzHMAAAWGujlyKqrNdrpdR8PrfWXl1dxV4NABBRO6YNISQ2OEVRxDZOSsk5r6qqqqosy2J8KqViKYzfDsPAGMMYU0pj1zSdTlEE1ul0OhqNIhJHACaEKKXi3WJfqZQihMQwM8ZEuIjQ7Jzz3sdXRciHEM5mM0IIIWBvb8taGYJhDBECOMcQuqJIYv/ivc+yLJazqqpiV7ZarWJ5jc3oZrOJAea932w2VVXFy3RdJ6VESilrLYQwXiDP85jsEU845/GUMfRjY4IQMsZkWTaZTBBC4/E4djXe+8Visdls3nvvPYTQwcHBYrEwVlf1BsAAYIAIYIIWy3mSitV6GU2DMXbOKaUwxsaY6Jk4AhhjoouyLItPRjOlaWqM2Ww2sRlF8RV930spYzSPx+MkSWRfOzMIhhgBFIehq/p2A7zOMh6CEYL0fe1kTZFTssvKNEmEVSrnopU9NRS4uiyz1XJxf/1g3q/UJkyK/VbLdb3M00SrDmELqUfMybbWdYcQWq1WSZKUZQmARwiE4MoyB8CnqQjBAeDX62Vdb0ajIs/TPM/j3UIIJMuyvu+vr6+n02nf9xF3vfdZlpRlKaWMqBrNHxEwdpqEkA6FvmkZY1BJShLExGao62F1v774x//sn2R7+3/x/e+8/fgdltHzs5NfuPuzN3i2Ojd5uSUHnfNk3VxTKBpXG+yIlLGHa9s2y5KPAAchY0yEASnlaDTSWi8WC4wxAEgpFQsFilPF7u5uTNmIaPGWMXsghGmaWmtHoxHGOE5PxhjO+QiKqcgmVOQA2ro5ff++q1fTNPzDX/xZb2R33UCJb+0dzReXkm6+852vy7OaT/r/+Y3fHcTVyp6lBWuqvmpaPhZpmiZJkud5zIo4tRljIlrEOSlCXxyGqqqKNbttWxJPE/NVKRWP6L1Pko8+jBUgdove+zTNYhUPIQy6tdDVXr5/+r7IyOs/85nvv/nGatiYi/W9nY9dX3qEwzvn7852d9oH7bM//IP3z04uH5yQ5/z/vf5ae929cvTKxO8fzPbOry6wp3me13U9mUwIQW3bJkkSB5ssy2L5j5kZG++Dg4PNZhPTGv7Hf//bsVgMw+C9T5IkzhlCsFgEIITEy47mJuCk7SFiG98twkKMCcl0N/TldCKtebz8oAzMNwbkOejM9uzoZ179wnf+8hvf3bwdAux6mWXldDolhmo7VO3GefUcP5AIwrKgWZJv2MylYEAhwFBC2TijVJ4xxkVsgqy1TdPEbjqE4FyIl0yShMQKH0e+2ITFghDCR1W5KIp1HeymKsrsSXXqRTAMoBEr92dNdznemjkIZK9m0wMwr8eWbNnxe2g+6P7h6fH56npxuSnGJUFUys7ajAYQlHnp9gvL5fLJ9dxCJ1fHSSZujo42SzMT28GS0TAiAJMkUabTxiKEnk5IsfPDGBOC+76P+Eu2t7fPzs4ihFlrhRAQQmMM5zR2+YvFovVBmCHJCeP0Sf2YF+V0b385LI3SFkApZV3XtDe7s62Hmw+WKHhoWcqX7frq7LQw8CiZDXqwBOhVW4PG6BAWC0zY9PZ+Wy8Xx6fBsw9Nc7h1m41pQvLmgyXGmZQ6y9nQyPF4HPu0uDeYTCbOOQDQZDK5vr7OsoysVqs0TWPSIPRRyxAnHqXU5eXler2e7gqwz7918W3PKE1E3Vbdow5pp7SfTCZO6ZdfeOn6/Kx1UtyeVl6lg2eCTfN8Csgtlt5Nx0vLHp/Nk6IgBFmOgoOdbNtWBgQOD+6+++670z305+9890lxzB149uhjYJCpKL0DsWjEeiqEMMZIKUMIUuo4zUEIUbxirFadHLDAiIWhbx6/e/r4waN5tbZJtlid5QaUACrUaL1x9QYt+8Lln+TP/vSrPz6b7r316P0OyE71sukLmCzIYKxKRZLTHOPh8oN30MXq07df/tKP/cLHZ0dbl5357tvkgw+vTj50XUcR++m/9/NqPeyNtpSRxf7sw8vHV+TJOlwiFZDCTdVKY+u673vpPTDGeQ9iyUvTtOs6EmtebCSno3K12iznizwv6S6TbtjbsfPmXdMVS9JpSp1zbDYrZryrZDbe7id5vYc379RQ9zgdx6jt+x61Ho3BpWoe8GoXZNloPFhw0T4e3g+L9fm5Wsvt9No2o63Rpt9A3Z1+/QlhFFMy1NXF1aVS5qUfemn1aKPoQEmiKwUsKqaFMSouMhBCcdDrus45B3/r3/xGXAVYa50zpvdPzi4au6HPN5CFQIANBDGGLQSBYp6Wnu/e2O26rr1eJtBWdmiQAZz0m/727dvn5+cIoV02PeRbL2wdnF/df1cOLuen3byYt69MD0+Pz+++8OwfvvGndswV1KzIFpdXe9NpoLxpmvF4XBTFyfmTAFzORjfGB9toB7cipTlJaN+0cSLN8zy24rEDQrGETadTjDEuyuuqYiVKboGdmzsAoTzdFnBmO40QwRhmFI9Z0lXVxflD1V6GrvOdZJh5B4QQ8/m8bdssyx5eP1y0y4v59XfefHupOuOdrzohxAfz8/Ek686vPzk5eHV0JFpaLdpyNHMO9sMgkgRA2PX9OJuVYiK1anG9cItet1bZGDwhhKIo4jAda59SiiR5hjC+WswhhME2jPKlWU7uQQClw/1Vs0GMipRbY0bJbHdUXtXrru4UNLZk55v5eLQDPE0ohrBGiOW7B/PFRTnZwrvbjy0gLz1rsaUp25rtqF7ConzPq/V6JYMsg3/9R7/wte/8n94MlRlQb4/u7D158oTg1HlTTFg5Kq4u123e7924KSy/ml/mfOQ9UMZRnmRJEkLA2Pe9JLEjCiFwzgEge1vp+vz08viKJKgfpPbOdEMTMEVCG9AMslXNR5uOgLJpMRiFMZJBGhEI8kO9UnogLKHBiwEwifJ8fHW6nO5vPTLXFTAEhHI8Qi32IPzZ97+XF+NP3n31z/78z2++fHu+OGFlIgf17MGds8vHslNFub2/veWVDYFsTQ9k18ZMiIP104kFxZEgSRJCCNIYM9fq9SgdAYiKvBxn5TgpEGQfe/FFkWcyaJYQnjJMSTmegIBSnqm6NqvVMx0tT6ofpFv58aZ7fM6t9yV7c7g8ydxVBj5cXn7+9dcnSZaXpQegGI0wpWxSQMQuTuc//vmfWL13/fzomf2QvzDeuXp04irLNO0X63ZeD9UALW6qLs5MjDFCSJyc4roI/off+e3YkHrv14v1ylz0yYJvIUMCcD4Y21T1ZLLfDn2APiDIPEaIyMExmgLnuceiNWPoC+NoXq69XYMujNOt6c1Vr67NHCLivVdNJwIyWnsC0jTVWsedlxokQcBbB1rCAcVAE6hPr04ePHgymOHus0c/8IM/vA9u9SuDEkQxicVBCEEQGoZBCME5J8HDrh8gck1Tn56fj57PCGejNN30NSHpoNy9519c9udlEHyps8rfKm+AJDlrmp5QbsGsKH7q5774R1/9w8c7pJov66q+e++Zv3r0cMnXr916AX/QZThbrhY4pB10FgVunR0MpwwAV2+qPM+VUhCivPCE8+ePXv7Lb3/vY594pQ59H0xH0Rvf/fqrL3+inBzYduSBKfiW8dInAwsZhEJrPR6X8Ld+899qra+uzzabdbupNnT1Q3/nla7akO3cWRA8FSJDjKpWwpU6YCUyuixLq51qh9WEtFUNCV7rtg3DOM+WV1fSqduo1Hl69/Docnn2CEqBqVpWs9nMwZBI21ttrfXG4oTGbk0IASjoO3fv4B6U/u3TdwmWXb9ORTavr0c561fhc6/9BKhDqFnGUoeHjBVxQA8hkOVyWVWVNn3XdZ/44mvp4Y7LaFitJdSBAkYT6wFnGZ0FncgrByzWTJpn8+KZ2ehX/ubrj45P3j07+97Z8dYGzkA+PTxQTt+/Ojs17v7jByYDd0g+SQrPcb7w1Pi3x85jKAddpply+umKerB2WkxVs3rx1r3c4PcvHgCBQI5KcTAa8VqfNGJ+8ujkQDwzErdGYLLu66Io4pqLxGnh+vq6qjbsYMuMR2tjwKgk0sQFFue8aleGBEjgZb3MixlG9o31WaLMn3zlSdu2t597/jiRm4K9Lyvfzh3yPIWX0G5NZ6ZersNwvlj3Q59AnHJKYAoQYoIrrQP6qDGTUqYYdZfXSToiwSTV+rMvvfLNx29d9os7W7fqqro9e+7y4VUySpftYqSnXjpRiDhakDj3xGUypfTywfHsOZFnXPWDx9hZqWTjncyS/eC1ksPO+IbsK+PNBjmTgOHs6nB352pxmXNUDY1q20QQREmj+nRry16vnvd8kKaHuM8mKBPOOcC5HroAoTLaGBWHxzRNFTCtkwDKP/jeNxwD4njzU3/77/7RV/7Xoj2FGBtvArGwZtNy6hGgeeKDib2p1pr0QwtR0Fo5b772+//j5371lxqHpR6AVnDQufW6tYvuylrddnWXJjxgFQJDbBQIOdodVD/vNYTBeaMppIyDAGgxap8sWFq8g3sCIcDIQWeGVgV3kCfqqh1vbTeQ2AEFTEHwAIQCJLtF2Vi1tr2Qjhby/uPvF/t7myf3HXUQ44KNZc51cCzjUBPZDUmSeIDycoyShFurIQoA+NTxL//u76nFCbAboFogW7NecNkr2YPgOKXWKAnaoZ0LHACHV9Vp28+x742qsLE5oWbondXI+nE5+smf/ZJJyAaHJXI1AQZC5OHF6poklGiTDvaf/8Kvvrj73CQf53mORTp3hmv0KbzzxZuv3kHj9uTk1la5PbtJUJ6ICQKpB+S1l18PhhuAnlKEm80GGaMpQ0IwADwjCdbWDt2malvsK+rrglwwyyDkGGecqbZV2hxNd25iMjbtHZDs1YGfV+VgwzCYvuMYQRgIowkXb3z9W6JSu2U2ShgXKM9YThGFUDvdNNWt/f3hwfEXX/sbqUg33RAGmQGUCdE6tVgvFutVXbX379/3HjgXtLKMicO93csHx4kn68VVbFI/2s0457TWnHOlqMNWSkmJSItp6BqRjFs0oIS61bzdVAiBQXZiDic7+bbTXK9xeigH4mF23Q6spFlRVLIjlK/qakoLFvi+49fz66RIcUKsbLjzwIZO9pxlj08evXl+nHyP/cov/fJ//c//pcs10Pac6o5auF5r30+2twcDPnfvmeqt71FB265yRt7dfjms7LQQ1trJZDIMQ1mW8Jf/0S9W1bppquVqrq0Hrl/o+U//8j/Q2Sz00puu6ZsAA8VkaFo9yP2bh91qBbUOKDjYYZJBSFzfYA63AV8loIMONH3GD6vuemu3qAZGMWEEaDXsbW+fPvrguZ2bqSHH7z3cgG7g1PH04ObRvaNnvvXdb0LsEEWjbAsSuNzMlVX75WHdVp3uO9ntsfErs89ISwXSAIvIpnrvCec8LmCEEHLdOoUOJzeXbz48up3W7WbjWpOBjOOhqxkjENKz82ORCU2NcXpKkwEHRz0AFGjd2q69ql574RMpdYuGHq+rAcki22NMwBBAICfHF8aj5bqu18M0y7jRm4zPsX5w9YA5cGN7//zq0ls4hMZ6RxGFGF6uz3d2t7PAqwrcEDdHaW57gyFhSRKlAnGfSTjnw9BRSjkGGrO+02q9xuJqu2SIkRO9VgYBEJSRVPBEOdd3guExI8QxBM0Ec7xUt9ju+3otRGbvX4/u7O0/9+I7i0fWGmv1ZmMm5WRrtnt2cobS/Bw5ejvvZU+0GJSkDqTGndfXGNDt7YOL8znPOcEkL6dpVjw6/uD8siEgZOlomu+0mxqJBEH6lPnNsowkSRL3mwB6ZOSqUVgkR8/cPDnkZFD15cU05WfL+e7+fsBwMNpwaILjFIxZur2kDffBm71nbj08uTpmdkt5er2ByI7ZDgAg55nWmiC6Xq/VYMpy/CNf+Nwf/eFXaGUyC1xSeAODs2mWntfLSTYLlBeTmfRKSbOo5whXRT4DToyylGBYLZud/NDnWTC+G4Y4T2OMEUFQMJ7wNBOlttgRjZTne6Oy0f18SVOx8QZy3Ot+vV5yjIIGs3Sak3Fr8NkEhcn4+NH828cPTpLeAXVBuw8PwIf7o/zuUa0UzsoBQOM8682RQp8d7T/8yrd/7Z/+a7F9k9zY0w5jQDAl2ruD6bbgPGDER/lsvM/GE56LEYN83qciUxDn+ezWjXsdxV4bgmCZp0WWGDUIRlDcznLOKaUHBzeFSOWgvvrVr8nQeeQCxJxku7u7cVXY9z1FxCqjB5UwsTM5WG16vjMGJX5KSe3v7w9Vc7B/gwnerDYFzXZA9pniNq/som7OF9e/99/++2de+SLoJ4xmAHEbiMccgYLTMWdl12oQOJSYBFbkW2SUJpBmA9zySbupOGVt21pr4+I1bm6ItTZKAhBCRtqLzSXO4K0Xdj3m5Za4PD/b29tTTk0mk67rKKVWIU5FlmVa68X1mQlKGeO0chBEXctisSBefPnLXy54ciPN9+Cu1f3Cm/lu+aC9QLssrB9tvmlHqDwZgpRQBT+alL0JJPiz1ZUDYCRwBlJj1HW3Gd3Z45U7cqNdlfqCGmNms1lMZedcURTDMBDvfRR7OOfKsvz0pz/tGzunT3g22VSL6dbEB8m5iKIW7z0gtG1bwfhqsTS21wAAAEuQnLXXxpioe8jK4mjvaPHWE6iAu3F4WSuU4ouhwoTkHla+tsJWbviRz//oV7/5x4Kk2hnrjO0kTQSE8HR5TiGEGPBxsr00N4u90ifEkU299hAg9P83yqvVand3F/76v/wXcauMMbZWIwwAcAB6NfJih79zch97HIK3MFjjciZa1WOLunqY7Iyrdhk8hAE645vNHDCmQ3jh3ot6DU0YLuanmCchEZN8bHtLRaZCcC5AhpTTEKPCz/q6Cgx4TrgjJGHWGyqolYGRwKFLGH7JHCRJIqWMVo4UVuTO47ERQoQxFveYkbwKwEFIEALmam2lgq1VpEOASm8hQ5u+ZzSdTCYYUITQfjpdXS+20nHCGbt5KLF78/S9d++/lbmdZ+/deu+Dv5rsJM281kspIJ/t5JkQIJt2UiFoCRGDBYMgWTE21oHcW613tyfB+ztTgZT2dW02vRMuRn8krfM8j0Rz3NNEtoR476PICSHkvElE1nUD4XzCp7pVeE7HR9O1bstxsmnX2lvdttiirdmOhWbRbRqOr67PjvYOwbo6XVzgScqcSwh/+63vzybTvu2mgdm2R15XnfaErPApS3NAuFSGiSRN8+7yansyHVk0LbemIr84OeXCZ5wBnOMsHyCMtEPcBUfy1nsfu26tNYkUnVIqrsUHOfR9nyQZo6JRA7HZve1PXCzOZrcOl3JjPAaEA2MghOv1GtBAEcnzcmuys1yuOro22wxYl8NM9kMqxHhc5gE8WZ0prRKeWGBsQCNEgNdDE7Z3dozSuLv++M7ejJq9dDdL6NmjRx+fTBDhJ+cXPEkxJQlmQojImsfdRexYo34ikl0ksrxxd0lp4oF3zkLuIaaAQgjADtpzHdVP1tbJOpU4FQ5DkQllbObQwY2js/NL1dSzjNarFgYmJnR688abf/UXV9XFZGs78wIGr7o+hEApraohSachYLsWL9+6ecD3rbLKO2W01MphPjgIrJlOZpHDbppOSg0ACgEySpWUGCFrrTJaCMEoWdYVirq/uLGMKkprbVQ8GmOizkltmmdv3hmB4kc+/rmd8T4GWBkLMCCT9Hjx5PHlA8bhQY2++MzLX/r8jzVVvV6uOKEMM9lr4HCRjjFgDCcoUIbTaZbd3Z3d3RkXfSJ8kohyaFWUNXDOn6r3/prqBUoN3ltjVOT2YyGKFnHOTadT+Bv/6teyLIsMb1QIxJ/xoajaYJiG4AG0NlhTQOP0RjWAg6Zv5vWCClhdryR0O9nW5mQxLYpNO1ytLm/cOkyLrdPj08h2Rt3R9miyXU4Pp7vYoxGaOAcQx4QhKXWWZVVV5XlutI3sPYSQUNx1XZIkXdelIo0aOYyxlEMkSzHGJMqqoq6j73sIYRSfWGsjcwEhNMAhSgVJsNRJBQeLOKdCiIEMGaJrU5HdQ7dZIRXkoM/h2ne6LIpqtbGWCZaSlMheG+V2d3d3ptPn9p87Gh0+fu+kSxUXiXOSOeg9HIYhy7LIIsQdXtM02gyM0aLI4uI+amiiRCgKswgh8D/9zr8LIbRtG0KItEpk/KMfIhpENUMsKwF+5CvvPUU8kND6zenqSaXXCMJMFA8fPOyVHKWlUgqWnAXmgrfef/ZTr1Htp/yG9z6ux6O3oy0pJhjjaGZldHwGIXR4cLPv+0gxQwCipEFrHS+cpmmapoRzrrUuiiJW76f6nSiBiox1JIljV+JBSNM0vhcRACGaki2c0IWeX1yeuzLc3r29WF8/88ydt97+PgkMOrs/27l3+/nMJgwTqXU0VjRTzENKKcUkNkKxG9VaT6fTSHPGq3rvszSNARJCiKpG7/18Pv9I3yqljIAVCeDImUbFWhSLIYScc5TSXg7xbpzztq2N0gzznGXZdnJ7/3YvuyfnT8ow7VZdIvCN3ckLdz4lm4FrGjSwBEEcoi4sym6i8JBS2jVtjNvIMkeJq9a67wYIYWxJvXMRY+JOLF4mTVMSbRxp9phG8ZMYS2VZxh0UISR+m+d5/MUYM0rLyldZmTR9ExCkkI3H05TkGGQfzt/6+b//A1Yrv5kBX5VpUpNBomCqnnMeZ/koIIo2jjkQ9/JPYyF+FeVFXdcRjGMliAgWZUTGGBQPFKmhKPCJPHbMnigPjGEW6yVBmCCsBskpG7RK0xwGwlEqQpKgpG87j5xBzdHu7UffXbzzrWPn5sUUbOTSW+sbWWT5ztZ2wkVbNwkXwAejNAJQCBGVIBhjgjCnbDqewACKPGeUaqXQX+upYyVwzgkhok9IFNR3XRdFfzFsYrhHx0Wo/mijRkh0V9wxRu9HK8YKGrFYDs45Myp30qREUGhlDw+OqqrzmWeEnp6e7uzsxDSLMRl1ktPpVGsdO3xKaVVVWmsIEWPso1jy/qPFPSHxYFLKLMtI/Pt4+ljPo6fi9iYS8bGaMMaiADsKweKu96nSKb4hQhklAgCQJImx3NlAcHJxcQUhRIgMXR+1xRFqItcfYyHKB6M5opgqyzLOxXw+F0LEJIyFImq/owrUGEMicR3VZU+NGpM7yrM+0oUTAgCItEUE4pj0T0XNESXinQOQBDMfECG46wdCghBpCC6EIKJoE4CovI3NQRzKdnZ2mqaJNo5Uv/d+vV6Px+No767r4kliRHxUBAn5f/BJWh83GhM3AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<PIL.Image.Image image mode=RGB size=65x65 at 0x1F2F127B760>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_image = image.load_img(\"archive/test/test/AppleCedarRust2.JPG\", target_size = (65,65))\n",
    "test_image\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_image =test_image = image.img_to_array(test_image)\n",
    "test_image=test_image/255\n",
    "test_image = np.expand_dims(test_image, axis = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "in user code:\n\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\training.py\", line 2041, in predict_function  *\n        return step_function(self, iterator)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\training.py\", line 2027, in step_function  **\n        outputs = model.distribute_strategy.run(run_step, args=(data,))\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\training.py\", line 2015, in run_step  **\n        outputs = model.predict_step(data)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\training.py\", line 1983, in predict_step\n        return self(x, training=False)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\utils\\traceback_utils.py\", line 70, in error_handler\n        raise e.with_traceback(filtered_tb) from None\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\input_spec.py\", line 295, in assert_input_compatibility\n        raise ValueError(\n\n    ValueError: Input 0 of layer \"sequential\" is incompatible with the layer: expected shape=(None, 64, 64, 3), found shape=(None, 65, 65, 3)\n",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Input \u001b[1;32mIn [10]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[43mcnn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpredict\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtest_image\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m      2\u001b[0m result\n",
      "File \u001b[1;32m~\\anaconda3\\lib\\site-packages\\keras\\utils\\traceback_utils.py:70\u001b[0m, in \u001b[0;36mfilter_traceback.<locals>.error_handler\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m     67\u001b[0m     filtered_tb \u001b[38;5;241m=\u001b[39m _process_traceback_frames(e\u001b[38;5;241m.\u001b[39m__traceback__)\n\u001b[0;32m     68\u001b[0m     \u001b[38;5;66;03m# To get the full stack trace, call:\u001b[39;00m\n\u001b[0;32m     69\u001b[0m     \u001b[38;5;66;03m# `tf.debugging.disable_traceback_filtering()`\u001b[39;00m\n\u001b[1;32m---> 70\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m e\u001b[38;5;241m.\u001b[39mwith_traceback(filtered_tb) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;28mNone\u001b[39m\n\u001b[0;32m     71\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[0;32m     72\u001b[0m     \u001b[38;5;28;01mdel\u001b[39;00m filtered_tb\n",
      "File \u001b[1;32m~\\AppData\\Local\\Temp\\__autograph_generated_file8nss73ka.py:15\u001b[0m, in \u001b[0;36mouter_factory.<locals>.inner_factory.<locals>.tf__predict_function\u001b[1;34m(iterator)\u001b[0m\n\u001b[0;32m     13\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m     14\u001b[0m     do_return \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m---> 15\u001b[0m     retval_ \u001b[38;5;241m=\u001b[39m ag__\u001b[38;5;241m.\u001b[39mconverted_call(ag__\u001b[38;5;241m.\u001b[39mld(step_function), (ag__\u001b[38;5;241m.\u001b[39mld(\u001b[38;5;28mself\u001b[39m), ag__\u001b[38;5;241m.\u001b[39mld(iterator)), \u001b[38;5;28;01mNone\u001b[39;00m, fscope)\n\u001b[0;32m     16\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m:\n\u001b[0;32m     17\u001b[0m     do_return \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n",
      "\u001b[1;31mValueError\u001b[0m: in user code:\n\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\training.py\", line 2041, in predict_function  *\n        return step_function(self, iterator)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\training.py\", line 2027, in step_function  **\n        outputs = model.distribute_strategy.run(run_step, args=(data,))\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\training.py\", line 2015, in run_step  **\n        outputs = model.predict_step(data)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\training.py\", line 1983, in predict_step\n        return self(x, training=False)\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\utils\\traceback_utils.py\", line 70, in error_handler\n        raise e.with_traceback(filtered_tb) from None\n    File \"C:\\Users\\ektap\\anaconda3\\lib\\site-packages\\keras\\engine\\input_spec.py\", line 295, in assert_input_compatibility\n        raise ValueError(\n\n    ValueError: Input 0 of layer \"sequential\" is incompatible with the layer: expected shape=(None, 64, 64, 3), found shape=(None, 65, 65, 3)\n"
     ]
    }
   ],
   "source": [
    "result = cnn.predict(test_image)\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
